geode-core/src/main/java/org/apache/geode/internal/cache/wan/AbstractGatewaySender.java

```diff
@@ -1118,6 +1115,24 @@ public void distribute(EnumListenerEvent operation, EntryEventImpl event,
     }
   }
 
+  private void dropEvent(EntryEventImpl event) {
+    final boolean isDebugEnabled = logger.isDebugEnabled();
+    if (this.eventProcessor != null) {
+      this.eventProcessor.registerEventDroppedInPrimaryQueue(event);
+    } else {
+      // Add empty event so that in case the event stays for long in
```

Comment:
the event is already a cloned. No need to clone again here. 

The if-else will reduce some chance to save into tmpDroppedEvents. But if the sender is purposely shutdown, the events will still be added into tmpDroppedEvents. This is your original problem. I don't think your current version will resolve this problem. Can you add a test case to prove?

Reply:
The newly cloned event does not contain the same informacion as the cloned event. Just the minimal needed to drain the queue of the secondary. This is just an optimization so, in case events need to be stored in tmpDroppedEvents, they take less space (for example, the value is not stored). With it, the memory for the original clone could be freed by the garbage collector.

With regards to the if-else you mention, when the gateway sender is started, an instance of an `AbstractGatewaySenderEventProcessor` is created and stored in the `eventProcessor member`. If the sender is stopped, the eventProcessor is not deleted. Therefore, events will not be stored in tmpDroppedEvents but will be sent to the other gateway senders by means of the call to `eventProcessor.registerEventDroppedInPrimaryQueue(event)`.
The only case in which events could be stored in tmpDroppedEvents is when the sender is created with `manualStart` set to true while the sender is started.

In the last two test cases added `stoppedSenderShouldNotAddEventsToTmpDroppedEventsButStillDrainQueuesWhenStarted` (line 581) and `stoppedPrimarySenderShouldNotAddEventsToTmpDroppedEventsButStillDrainQueuesWhenStarted` (line 655) you can see that events are not added to tmpDroppedEvents and prove that the solution solves the problem.

Reply:
I see. That makes sense. But please don't clone a new event again. The cloneEvent is created the same way.

Reply:
I reviewed the my change history, I was using 
 if (this.eventProcessor != null) {
      this.eventProcessor.registerEventDroppedInPrimaryQueue(event);
    } 
in version one. 

Then I found a bug that a newly started sender has a race to miss some event. So I change to use tmpDroppedEvents. However, we need to keep both.

Reply:
I am ok with not cloning again the event. Anyway, if you look at the cloned event it would contain the newValue while the newly cloned event will not. If the value is big, tmpDroppedEvents will take much more memory.

Apart from the above, do you agree with this proposal to be implemented in a pull request?

Reply:
yes, I think it's good.

Reply:
Thanks. I have written a JIRA https://issues.apache.org/jira/browse/GEODE-8491 and here is the PR: https://github.com/apache/geode/pull/5509